nf_core/pipelines/download/download.py

```diff
@@ -0,0 +1,717 @@
+                log.warning("Running 'nextflow inspect' failed with the following error")
+                raise DownloadError(e)
+
+            except KeyError as e:
+                log.warning("Failed to parse output of 'nextflow inspect' to extract containers")
+                raise DownloadError(e)
+
+    def gather_registries(self, workflow_directory: str) -> None:
+        """Fetch the registries from the pipeline config and CLI arguments and store them in a set.
+        This is needed to symlink downloaded container images so Nextflow will find them.
+        """
+
+        # should exist, because find_container_images() is always called before
+        if not self.nf_config:
+            self.nf_config = nf_core.utils.fetch_wf_config(Path(workflow_directory))
+
+        # Select registries defined in pipeline config
+        configured_registries = [
+            "apptainer.registry",
+            "docker.registry",
+            "podman.registry",
+            "singularity.registry",
+        ]
+
+        for registry in configured_registries:
+            if registry in self.nf_config:
+                self.registry_set.add(self.nf_config[registry])
+
+        # add depot.galaxyproject.org to the set, because it is the default registry for singularity hardcoded in modules
+        self.registry_set.add("depot.galaxyproject.org/singularity")
+
+        # add community.wave.seqera.io/library to the set to support the new Seqera Docker container registry
+        self.registry_set.add("community.wave.seqera.io/library")
+
+        # add chttps://community-cr-prod.seqera.io/docker/registry/v2/ to the set to support the new Seqera Singularity container registry
+        self.registry_set.add("community-cr-prod.seqera.io/docker/registry/v2")
+
+    def get_container_output_dir(self) -> Path:
+        assert self.outdir is not None  # mypy
+        return self.outdir / f"{self.container_system}-images"
+
+    def download_container_images(self, current_revision: str = "") -> None:
+        """Loop through container names and download Singularity images"""
+
+        if len(self.containers) == 0:
+            log.info("No container names found in workflow")
+        else:
+            log.info(
+                f"Processing workflow revision {current_revision}, found {len(self.containers)} container image{'s' if len(self.containers) > 1 else ''} in total."
+            )
+            log.debug(f"Container names: {self.containers}")
+
+            out_path_dir = self.get_container_output_dir().absolute()
+
+            # Check that the directories exist
+            if not out_path_dir.is_dir():
+                log.debug(f"Output directory not found, creating: {out_path_dir}")
+                out_path_dir.mkdir()
+
+            if self.container_fetcher is not None:
+                self.container_fetcher.fetch_containers(
+                    self.containers,
+                    self.containers_remote,
+                )
```

Comment:
Should we raise a `DownloadError` if there are containers but no `ContainerFetcher`? Admittedly, I do not know how that should ever happen...